presto-hive/src/main/java/io/prestosql/plugin/hive/parquet/ParquetPageSourceFactory.java

```diff
@@ -342,4 +352,30 @@ public static ReaderPageSourceWithProjections createPageSource(
         }
         return null;
     }
+
+    private static Map<List<String>, RichColumnDescriptor> getDescriptorsUsingIndex(List<HiveColumnHandle> hiveColumnHandles, MessageType fileSchema)
+    {
+        Map<List<String>, RichColumnDescriptor> descriptorsByName = new HashMap<>();
+        Optional<ReaderProjections> readerProjections = projectBaseColumns(hiveColumnHandles);
+        List<HiveColumnHandle> baseColumns = readerProjections.map(ReaderProjections::getReaderColumns).orElse(hiveColumnHandles);
+
+        List<Optional<org.apache.parquet.schema.Type>> parquetFields = baseColumns.stream()
+                .map(column -> getParquetType(column, fileSchema, true))
+                .map(Optional::ofNullable)
+                .collect(toImmutableList());
+
+        for (int columnIndex = 0; columnIndex < baseColumns.size(); columnIndex++) {
+            HiveColumnHandle column = baseColumns.get(columnIndex);
+            if (column.getHiveType().getCategory() != PRIMITIVE) {
+                continue;
+            }
+
+            parquetFields.get(columnIndex).ifPresent(parquetField -> {
+                descriptorsByName.put(
+                        ImmutableList.of(fileSchema.getFields().get(column.getBaseHiveColumnIndex()).getName()),
+                        new RichColumnDescriptor(fileSchema.getColumnDescription(new String[] {parquetField.getName()}), parquetField.asPrimitiveType()));
```

Comment:
> `new String[] {parquetField.getName()}`

this seems to be correct for top-level fields; is it also correct for nested?

if a have a top level `col` but also nested `r.col`, will this get confused?

Reply:
It would be a problem, but non-base columns are filtered out of the effectivePredicate. We'll need to fix this if we want to add support for pushing down filters on nested columns. 

https://github.com/prestosql/presto/blob/master/presto-hive/src/main/java/io/prestosql/plugin/hive/parquet/ParquetPageSourceFactory.java#L166

Reply:
I think we will want to push down predicates on r.col.
Actually, if col is the only field being read, don't we do that already?

even if we don't, if we have a schema:
```
struct r {
  string col
}

string col
```

here we could see r.col (primitive), we would find ColumnDescription appropriate for col, a wrong one.
-- is that possible?

Reply:
I don't think It's possible yet, because we're only looking up columns from this map that are in the predicate, and nested ones are removed. 

If we have a query filtered just off of a nested column it is still filtered out right now. 

I can add a todo comment to remind us to address this when someone gets around to adding push down for nested columns, but I added an extra test with this case too.